static/js/compose_ui.js

```diff
@@ -67,11 +67,39 @@ export function smart_insert($textarea, syntax) {
     autosize_textarea($textarea);
 }
 
-export function insert_syntax_and_focus(syntax, $textarea = $("#compose-textarea")) {
+export function smart_insert_block($textarea, syntax) {
+    function is_newline(c) {
+        return c === "\n" || c === "\r";
+    }
+    const pos = $textarea.caret();
+    const before_str = $textarea.val().slice(0, pos);
+
+    // Replace any number of newlines at the end of the content before
+    // the insert with exactly two newlines.
+    if (pos > 0 && is_newline(before_str.slice(-1))) {
+        $textarea.val(before_str.trimEnd() + "\n\n");
+    }
```

Comment:
This incorrectly erases any content the user might have written after the current caret location.

And the logic isn’t right. The desired behavior is that we add just as many extra newlines as needed to separate the quote from the previous non-newline character (if any) by at least 2 newlines; this may be 0, 1, or 2 extra newlines.

Reply:
Thanks @andersk! I'll work on fixing those things.